src/controller/CHIPDiscovery.cpp

```diff
@@ -0,0 +1,336 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <stdarg.h>
+
+#include <cstdint>
+#include <string>
+
+#include <controller/CHIPDiscovery.h>
+#include <lib/mdns/minimal/core/Constants.h>
+
+namespace chip {
+namespace Controller {
+
+namespace {
+std::string ToString(const mdns::Minimal::BytesRange & bytes)
+{
+    std::string str;
+    str.reserve(bytes.Size() + 1);
+    str.assign(reinterpret_cast<const char *>(bytes.Start()), bytes.Size());
+    return str;
+}
+
+} // namespace
+
+void TxtRecordDelegateImpl::OnRecord(const mdns::Minimal::BytesRange & name, const mdns::Minimal::BytesRange & value)
+{
+    if (dnsSdInfo == nullptr)
+    {
+        return;
+    }
+    // Names are just 1 letter.
+    if (name.Size() < 1)
+    {
+        return;
+    }
+    std::string key = ToString(name);
+    std::string val = ToString(value);
+    if (key == "D")
+    {
+        // For some reason there is no standard string -> short converstion.
+        unsigned long temp = std::stoul(val);
+        if (temp < UINT16_MAX)
+        {
+            dnsSdInfo->longDiscriminator = static_cast<uint16_t>(temp);
+        }
+    }
+    else if (key == "VP")
```

Comment:
Can you look for the other keys defined in section 4.2.1 as well?

Reply:
Is it OK if I add this in another PR? So far I've only tested with long discriminator checked product vendor. I'd like to give the other keys and subs a proper testing.